colab/edge_detection.ipynb

```diff
@@ -246,38 +245,40 @@
       "source": [
         "#@markdown ### Backend Configuration\n",
         "\n",
-        "backend_choice = \"iree_vmvx (CPU)\" #@param [ \"iree_vmvx (CPU)\", \"iree_llvmaot (CPU)\", \"iree_vulkan (GPU/SwiftShader)\" ]\n",
+        "backend_choice = \"iree_llvmaot (CPU)\" #@param [ \"iree_vmvx (CPU)\", \"iree_llvmaot (CPU)\", \"iree_vulkan (GPU/SwiftShader)\" ]\n",
         "backend_choice = backend_choice.split(\" \")[0]\n",
         "backend = module_utils.BackendInfo(backend_choice)"
       ],
-      "execution_count": 7,
+      "execution_count": null,
       "outputs": []
     },
     {
       "cell_type": "code",
       "metadata": {
         "colab": {
           "base_uri": "https://localhost:8080/",
-          "height": 264
+          "height": 282
         },
         "id": "2OPBr0eGroE0",
-        "outputId": "a0cc5e49-2ef1-483d-eecc-81d5b29b3e1d"
+        "outputId": "fd30a9fb-c741-4f70-b688-6419820028ed"
       },
       "source": [
         "#@title Compile and Run the EdgeDetectionModule with IREE.\n",
         "module = backend.compile_from_class(EdgeDetectionModule)\n",
         "\n",
         "# Compute the edges using the compiled module and display the result.\n",
         "iree_edges = module.edge_detect_sobel_operator(image)\n",
-        "show_images(image, iree_edges)"
+        "print(iree_edges)\n",
+        "show_images(image, iree_edges.to_host())"
```

Comment:
@stellaraccident Should this `.to_host()` be needed, or did you had an implementation of `DeviceArray.reshape()` in mind as the fix?

If so, we should be able to fix all the Colab notebook failures by just updating the Python bindings.

Reply:
I checked with ```.to_host()``` from [here](https://github.com/google/iree/blob/main/bindings/python/iree/runtime/array_interop.py). by applying this it is internally converting to ```nparray``` and resolving ```reshape()```. I didn't have any explicit implementation on ```DeviceArray.reshape()```.

Reply:
Yeah, there are a few methods already implemented on DeviceArray like these:

https://github.com/google/iree/blob/c4820685ae814c09c946b1badae23b0ff090b31e/bindings/python/iree/runtime/array_interop.py#L84-L86

https://github.com/google/iree/blob/c4820685ae814c09c946b1badae23b0ff090b31e/bindings/python/iree/runtime/array_interop.py#L130-L138

so we could (for example) add an implementation of `reshape()` that forwards to `to_host().reshape()`

It depends on how much we want `DeviceArray` to behave like `nparray`, and if this particular usage is worth an implicit conversion, explicit conversion (this PR), or a custom implementation of `reshape()`.

Reply:
Can you try `np.reshape(array)`? Generally, the aliases directly on an array are just sugar for those top level functions. It is better to add DeviceArray attribute aliases that delegate back to the free-standing functions.

Right now, they are all just faulting back to the host and doing the operation in numpy. In the future, we can implement some of these to be more efficient and not require the host-readback. But that is for later.

Reply:
like this?

```python

def show_images(image, edges):
  fig, axs = plt.subplots(1, 2)

  axs[0].imshow(np.reshape(image, (128, 128)), cmap="gray")
  axs[0].set_title("Input image")
  axs[1].imshow(np.reshape(edges, (128, 128)), cmap="gray")
  # ...

iree_edges = module.edge_detect_sobel_operator(image)
show_images(image, iree_edges)

----> 6 show_images(image, iree_edges)

[<ipython-input-10-c6aee5cdbc30>](https://localhost:8080/#) in show_images(image, edges)
     18   axs[0].imshow(np.reshape(image, (128, 128)), cmap="gray")
     19   axs[0].set_title("Input image")
---> 20   axs[1].imshow(np.reshape(edges, (128, 128)), cmap="gray")
     21   axs[1].set_title("Output image")
     22 

<__array_function__ internals> in reshape(*args, **kwargs)

TypeError: no implementation found for 'numpy.reshape' on types that implement __array_function__: [<class 'iree.runtime.array_interop.DeviceArray'>]
```

Reply:
Ok, I sent https://github.com/google/iree/pull/8388 adding stubs (this particular `reshape()` _could_ have a more efficient implementation, but I'm implicitly using `ary.to_host()` for now).

If that PR looks good we can close this one.

Reply:
Yep, that PR fixed the notebook and we don't need a direct call to `reshape()` in the notebook now.